base/shell/explorer/traydeskbtn.cpp

```diff
@@ -0,0 +1,387 @@
+/*
+ * PROJECT:     ReactOS Explorer
+ * LICENSE:     LGPL-2.1-or-later (https://spdx.org/licenses/LGPL-2.1-or-later)
+ * PURPOSE:     Show Desktop tray button implementation
+ * COPYRIGHT:   Copyright 2006-2007 Thomas Weidenmueller <w3seek@reactos.org>
+ *              Copyright 2018-2022 Katayama Hirofumi MZ <katayama.hirofumi.mz@gmail.com>
+ *              Copyright 2023 Ethan Rodensky <splitwirez@gmail.com>
+ */
+
+#include "traydeskbtn.h"
+#include <limits.h>
+
+#define IDI_SHELL32_DESKTOP 35
+#define IDI_IMAGERES_DESKTOP 110
+
+#define SHOW_DESKTOP_TIMER_ID 999
+#define SHOW_DESKTOP_TIMER_INTERVAL 200
+
+CTrayShowDesktopButton::CTrayShowDesktopButton() :
+    m_nClickedTime(0),
+    m_inset({2, 2}),
+    m_icon(NULL),
+    m_highContrastMode(FALSE),
+    m_drawWithDedicatedBackground(FALSE),
+    m_bHovering(FALSE),
+    m_hWndTaskbar(NULL),
+    m_bPressed(FALSE),
+    IsHorizontal(FALSE)
+{
+}
+
+INT CTrayShowDesktopButton::WidthOrHeight() const
+{
+    if (IsThemeActive() && !m_highContrastMode)
+    {
+        if (m_drawWithDedicatedBackground)
+        {
+            if (GetSystemMetrics(SM_TABLETPC))
+            {
+                //TODO: DPI scaling - return logical-to-physical conversion of 24, not fixed value
+                return 24;
+            }
+            else
+                return 15;
+        }
+        else
+        {
+            INT CurMargin = IsHorizontal
+                ? (m_ContentMargins.cxLeftWidth + m_ContentMargins.cxRightWidth)
+                : (m_ContentMargins.cyTopHeight + m_ContentMargins.cyBottomHeight);
+            return max(16 + CurMargin, 18) + 6;
+        }
+    }
+    else
+    {
+        return max(2 * GetSystemMetrics(SM_CXBORDER) + GetSystemMetrics(SM_CXSMICON),
+                   2 * GetSystemMetrics(SM_CYBORDER) + GetSystemMetrics(SM_CYSMICON));
+    }
+}
+
+HRESULT CTrayShowDesktopButton::DoCreate(HWND hwndParent)
+{
+    const DWORD style = WS_CHILD | WS_VISIBLE | WS_CLIPSIBLINGS | BS_DEFPUSHBUTTON;
+    Create(hwndParent, NULL, NULL, style);
+
+    if (!m_hWnd)
+        return E_FAIL;
+
+    // Get desktop icon
+    bool bIconRetrievalFailed = ExtractIconExW(L"imageres.dll", -IDI_IMAGERES_DESKTOP, NULL, &m_icon, 1) == UINT_MAX;
+    if (bIconRetrievalFailed || !m_icon)
+        ExtractIconExW(L"shell32.dll", -IDI_SHELL32_DESKTOP, NULL, &m_icon, 1);
+
+    // Get appropriate size at which to display desktop icon
+    m_szIcon.cx = GetSystemMetrics(SM_CXSMICON);
+    m_szIcon.cy = GetSystemMetrics(SM_CYSMICON);
+
+    // Prep visual style
+    EnsureWindowTheme(TRUE);
+
+    // Get HWND of Taskbar
+    auto parent = GetParent();
+    if (!parent)
+        return E_FAIL;
+    parent = parent.GetParent();
+    if (!parent)
+        return E_FAIL;
+    m_hWndTaskbar = parent.m_hWnd;
```

Comment:
```suggestion
    HWND hWndParent = ::GetParent(m_hWnd);
    if (hWndParent)
        hWndParent = ::GetParent(hWndParent);
    if (!hWndParent)
        return E_FAIL;
    m_hWndTaskbar = hWndParent;
```
Note also that it is possible to directly do: `GetParent(GetParent(...))`
if the innermost call returns NULL, then the other call will also return NULL, and the previous code could be simplified further to:
```suggestion
    m_hWndTaskbar = ::GetParent(::GetParent(m_hWnd));
    if (!m_hWndTaskbar)
        return E_FAIL;
```

_**In addition...**_ Don't we already know the first parent (i.e. what the innermost `GetParent()` call would return)? Shouldn't it be the `hwndParent` parameter that is given to that `DoCreate()` function?

Reply:
Good catch - I've cleaned that whole area up now :+1: 

Is this now resolved to your satisfaction?

Reply:
looks great :+1: